pkg/planner/core/memtable_infoschema_extractor.go

```diff
@@ -0,0 +1,493 @@
+	_indexName       = "index_name"
+	_schemaName      = "schema_name"
+)
+
+var extractableColumns = map[string][]string{
+	// See infoschema.tablesCols for full columns.
+	// Used by InfoSchemaTablesExtractor and setDataFromTables.
+	infoschema.TableTables: {
+		_tableSchema, _tableName, _tidbTableID,
+	},
+	// See infoschema.partitionsCols for full columns.
+	// Used by InfoSchemaPartitionsExtractor and setDataFromPartitions.
+	infoschema.TablePartitions: {
+		_tableSchema, _tableName, _tidbPartitionID,
+		_partitionName,
+	},
+	// See infoschema.statisticsCols for full columns.
+	// Used by InfoSchemaStatisticsExtractor and setDataForStatistics.
+	infoschema.TableStatistics: {
+		_tableSchema, _tableName,
+		_indexName,
+	},
+	// See infoschema.schemataCols for full columns.
+	// Used by InfoSchemaSchemataExtractor and setDataFromSchemata.
+	infoschema.TableSchemata: {
+		_schemaName,
+	},
+}
+
+// InfoSchemaBaseExtractor is used to extract infoSchema tables related predicates.
+type InfoSchemaBaseExtractor struct {
+	extractHelper
+	// SkipRequest means the where clause always false, we don't need to request any component
+	SkipRequest   bool
+	ColPredicates map[string]set.StringSet
+	// columns occurs in predicate will be extracted.
+	colNames []string
+}
+
+func (e *InfoSchemaBaseExtractor) initExtractableColNames(systableName string) {
+	cols, ok := extractableColumns[systableName]
+	if ok {
+		e.colNames = cols
+	} else {
+		// TODO: remove this after all system tables are supported.
+		e.colNames = []string{
+			"table_schema",
+			"constraint_schema",
+			"table_name",
+			"constraint_name",
+			"sequence_schema",
+			"sequence_name",
+			"partition_name",
+			"schema_name",
+			"index_name",
+			"tidb_table_id",
+		}
+	}
+}
+
+// SetExtractColNames sets the columns that need to be extracted.
+func (e *InfoSchemaBaseExtractor) SetExtractColNames(colNames ...string) {
+	e.colNames = colNames
+}
+
+// Extract implements the MemTablePredicateExtractor Extract interface
+func (e *InfoSchemaBaseExtractor) Extract(
+	ctx base.PlanContext,
+	schema *expression.Schema,
+	names []*types.FieldName,
+	predicates []expression.Expression,
+) (remained []expression.Expression) {
+	var resultSet, resultSet1 set.StringSet
+	e.ColPredicates = make(map[string]set.StringSet)
+	remained = predicates
+	for _, colName := range e.colNames {
+		remained, e.SkipRequest, resultSet = e.extractColWithLower(ctx, schema, names, remained, colName)
+		if e.SkipRequest {
+			break
+		}
+		remained, e.SkipRequest, resultSet1 = e.extractCol(ctx, schema, names, remained, colName, true)
+		if e.SkipRequest {
+			break
+		}
+		for elt := range resultSet1 {
+			resultSet.Insert(elt)
+		}
+		if len(resultSet) == 0 {
+			continue
+		}
+		e.ColPredicates[colName] = resultSet
+	}
+	return remained
+}
+
+// ExplainInfo implements base.MemTablePredicateExtractor interface.
+func (e *InfoSchemaBaseExtractor) ExplainInfo(_ base.PhysicalPlan) string {
```

Comment:
I'm not sure what's the expected behaviour of `ExplainInfo` because the comments are poor. Should it be the used filter for a query rather than all filters?

Reply:
Yes. For unused filters, they will be displayed as another operator. For example:
```
mysql> explain select * from information_schema.tables where (table_name = 't2' and tidb_table_id = 3);
+----------------+----------+------+---------------+----------------------------------------+
| id             | estRows  | task | access object | operator info                          |
+----------------+----------+------+---------------+----------------------------------------+
| MemTableScan_5 | 10000.00 | root | table:TABLES  | table_name:["t2"], tidb_table_id:["3"] |
+----------------+----------+------+---------------+----------------------------------------+
1 row in set (0.00 sec)

mysql> explain select * from information_schema.tables where table_name = 't1' or (table_name = 't2' and tidb_table_id = 3);
+----------------------+----------+------+---------------+-------------------------------------------------------------------+
| id                   | estRows  | task | access object | operator info                                                     |
+----------------------+----------+------+---------------+-------------------------------------------------------------------+
| Selection_5          | 8000.00  | root |               | or(eq(Column#3, "t1"), and(eq(Column#3, "t2"), eq(Column#22, 3))) |
| └─MemTableScan_6     | 10000.00 | root | table:TABLES  |                                                                   |
+----------------------+----------+------+---------------+-------------------------------------------------------------------+
2 rows in set (0.00 sec)
```